contracts/strategies/auraBAL/Strategy.sol

```diff
@@ -0,0 +1,202 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.9;
+
+import "@openzeppelin/contracts/access/Ownable.sol";
+import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import "@openzeppelin/contracts/token/ERC20/ERC20.sol";
+import "../../../interfaces/IGenericVault.sol";
+import "../../../interfaces/IStrategy.sol";
+import "../../../interfaces/IRewards.sol";
+import "../../../interfaces/balancer/IRewardHandler.sol";
+import "./StrategyBase.sol";
+
+contract AuraBalStrategy is Ownable, AuraBalStrategyBase {
+    using SafeERC20 for IERC20;
+
+    address public immutable vault;
+    address[] public rewardTokens;
+    mapping(address => address) public rewardHandlers;
+
+    uint256 public constant FEE_DENOMINATOR = 10000;
+
+    constructor(address _vault) {
+        vault = _vault;
+    }
+
+    /// @notice Set approvals for the contracts used when swapping & staking
+    function setApprovals() external {
+        IERC20(AURABAL_TOKEN).safeApprove(AURABAL_STAKING, 0);
+        IERC20(AURABAL_TOKEN).safeApprove(AURABAL_STAKING, type(uint256).max);
+        IERC20(BAL_TOKEN).safeApprove(BAL_VAULT, 0);
+        IERC20(BAL_TOKEN).safeApprove(BAL_VAULT, type(uint256).max);
+        IERC20(WETH_TOKEN).safeApprove(BAL_VAULT, 0);
+        IERC20(WETH_TOKEN).safeApprove(BAL_VAULT, type(uint256).max);
+        IERC20(BAL_ETH_POOL_TOKEN).safeApprove(AURABAL_PT_DEPOSIT, 0);
+        IERC20(BAL_ETH_POOL_TOKEN).safeApprove(
+            AURABAL_PT_DEPOSIT,
+            type(uint256).max
+        );
+        IERC20(BAL_ETH_POOL_TOKEN).safeApprove(BAL_VAULT, 0);
+        IERC20(BAL_ETH_POOL_TOKEN).safeApprove(BAL_VAULT, type(uint256).max);
+    }
+
+    /// @notice update the token to handler mapping
+    function _updateRewardToken(address _token, address _handler) internal {
+        rewardHandlers[_token] = _handler;
+    }
+
+    /// @notice Add a reward token and its handler
+    /// @dev For tokens that should not be swapped (i.e. BAL rewards)
+    ///      use address as zero handler
+    /// @param _token the reward token to add
+    /// @param _handler address of the contract that will sell for BAL or ETH
+    function addRewardToken(address _token, address _handler)
+        external
+        onlyOwner
+    {
+        rewardTokens.push(_token);
+        _updateRewardToken(_token, _handler);
+    }
+
+    /// @notice Update the handler of a reward token
+    /// @dev Used to update a handler or retire a token (set handler to address 0)
+    /// @param _token the reward token to add
+    /// @param _handler address of the contract that will sell for BAL or ETH
+    function updateRewardToken(address _token, address _handler)
+        external
+        onlyOwner
+    {
+        _updateRewardToken(_token, _handler);
+    }
+
+    /// @notice returns the number of reward tokens
+    /// @return the number of reward tokens
+    function totalRewardTokens() external view returns (uint256) {
```

Comment:
I think you can remove this function from `Strategy.sol` since you're not using it aside from some tests, but then you can just call `rewardTokens.length` directly because `rewardTokens` is a public array.

Reply:
the default solidity getters for arrays do not return the whole array, you have to pass an index. so i'd either have to iterate over the whole array with single calls in my test to assess its length or create a special getter function for it.